File: days/day07.py
from etc.utils import file_to_lines
import re

RE_OUTER = re.compile(r"(.*) bags contain")
RE_INNER = re.compile(r"(\d+) (.*?) bags?")

# Parsing
lines = file_to_lines("input/day07.txt")
data_tree = {}

def init_color(color):
    if color not in data_tree:
        data_tree[color] = {"contained_by": set(), "contains": []}

for line in lines:
    outer_color = RE_OUTER.match(line).group(1)  # Will always match
    init_color(outer_color)

    for m in RE_INNER.finditer(line):
        num, inner_color = int(m.group(1)), m.group(2)
        init_color(inner_color)
        data_tree[outer_color]["contains"].append((num, inner_color))
        data_tree[inner_color]["contained_by"].add(outer_color)

# Part 1
visited_parents = set()

def visit_parents(color):
    parents_to_visit = [x for x in data_tree[color]["contained_by"] if x not in visited_parents]
    for p in parents_to_visit:
        visited_parents.add(p)
        visit_parents(p)

visit_parents("shiny gold")
print("Part 1:", len(visited_parents))

# Part 2
def count_inner_bags(color):
    res = 1
    contains = data_tree[color]["contains"]
    if not contains: return res

    for num, inner_color in contains:
        res += num * count_inner_bags(inner_color)

    return res

num_inner_bags = count_inner_bags("shiny gold") - 1
print("Part 2:", num_inner_bags)

File: rust/src/days/day07.rs
use std::io::{BufReader, BufRead};
use std::fs::File;
use std::time::Instant;
use std::collections::HashSet;
use petgraph::graphmap::DiGraphMap;
use petgraph::Direction;
use regex::Regex;

///////////////////////////////////////////////////////////////////////////////

pub fn run() {
    let time = Instant::now();

    let f = BufReader::new(File::open("../input/day07.txt").unwrap());
    let lines: Vec<String> = f.lines().collect::<Result<_, _>>().unwrap();
    let graph = create_graph(&lines);

    let sol_part_1 = compute_part_1(&graph, "shiny gold");
    let sol_part_2 = compute_part_2(&graph, "shiny gold") - 1; // Exclude the shiny gold bag itself

    let elapsed_ms = time.elapsed().as_nanos() as f64 / 1_000_000.0;
    println!("Part 1: {}", sol_part_1);
    println!("Part 2: {}", sol_part_2);
    println!("Elapsed: {:.3} ms", elapsed_ms);
}

///////////////////////////////////////////////////////////////////////////////

fn compute_part_1(graph: &DiGraphMap<&str, usize>, initial_node: &str) -> usize {
    let mut visited_nodes: HashSet<&str> = HashSet::new();
    let mut nodes_to_visit: Vec<&str> = graph
        .neighbors_directed(initial_node, Direction::Incoming)
        .collect();

    while !nodes_to_visit.is_empty() {
        let node = nodes_to_visit.pop().unwrap();
        visited_nodes.insert(node);
        nodes_to_visit.extend(graph
            .neighbors_directed(node, Direction::Incoming)
            .filter(|n| !visited_nodes.contains(n))
        );
    }

    return visited_nodes.len();
}

fn compute_part_2(graph: &DiGraphMap<&str, usize>, initial_node: &str) -> usize {
    let mut res = 1;

    for node in graph.neighbors_directed(initial_node, Direction::Outgoing) {
        res += graph.edge_weight(initial_node, node).unwrap() * compute_part_2(graph, node);
    }

    return res;
}

fn create_graph(lines: &[String]) -> DiGraphMap<&str, usize> {
    let re_outer = Regex::new(r"(.*) bags contain").unwrap();
    let re_inner = Regex::new(r"(\d+) (.*?) bags?").unwrap();

    let mut graph = DiGraphMap::new();

    for line in lines {
        let outer_color = re_outer.captures(&line).unwrap().get(1).unwrap().as_str();
        for c in re_inner.captures_iter(&line) {
            let num: usize = c[1].parse().unwrap();
            let inner_color = c.get(2).unwrap().as_str();
            graph.add_edge(outer_color, inner_color, num);
        }
    }

    return graph;
}
